File: go/ingest/queue/backpressure.go
// SPDX-License-Identifier: Apache-2.0

package queue

import (
	"context"
	"sync/atomic"
)

// defaultMaxQueueDepth is the backpressure threshold per tenant.
// Modelled on RabbitMQ/SQS patterns where 1000 pending items per
// tenant signals the system should stop accepting new work.
const defaultMaxQueueDepth int64 = 1000

// BackpressureChecker evaluates whether the queue has exceeded its
// capacity threshold. When backpressured, producers should stop
// enqueuing new jobs until depth drops below the limit.
type BackpressureChecker struct {
	adapter       Adapter
	maxDepth      int64
	backpressured atomic.Bool
	lastDepth     atomic.Int64
}

// NewBackpressureChecker creates a checker with the given depth limit.
// A maxDepth of zero or negative falls back to defaultMaxQueueDepth.
func NewBackpressureChecker(adapter Adapter, maxDepth int64) *BackpressureChecker {
	safeMax := maxDepth
	if safeMax <= 0 {
		safeMax = defaultMaxQueueDepth
	}
	return &BackpressureChecker{
		adapter:  adapter,
		maxDepth: safeMax,
	}
}

// Check queries the adapter for current pending job count using
// CountByStatus and updates the backpressure state. The brainID
// parameter scopes the count; pass an empty string for global depth.
func (bc *BackpressureChecker) Check(ctx context.Context, brainID string) (bool, error) {
	counts, err := bc.adapter.CountByStatus(ctx, brainID)
	if err != nil {
		return bc.backpressured.Load(), err
	}
	depth := int64(counts[StatusPending])
	bc.lastDepth.Store(depth)
	pressured := depth >= bc.maxDepth
	bc.backpressured.Store(pressured)
	return pressured, nil
}

// IsBackpressured returns the last known backpressure state without
// querying the adapter. This is safe to call from hot paths where a
// stale value is acceptable.
func (bc *BackpressureChecker) IsBackpressured() bool {
	return bc.backpressured.Load()
}

// MaxDepth returns the configured threshold.
func (bc *BackpressureChecker) MaxDepth() int64 {
	return bc.maxDepth
}

// LastDepth returns the last observed queue depth from the most recent
// Check call. Returns 0 before the first check.
func (bc *BackpressureChecker) LastDepth() int64 {
	return bc.lastDepth.Load()
}

File: sdks/ts/memory/src/ingest/queue/backpressure.ts
// SPDX-License-Identifier: Apache-2.0

/**
 * Backpressure detection for the ingestion queue. Monitors pending job
 * count via countByStatus and signals when producers should stop
 * enqueuing new jobs. Follows the RabbitMQ/SQS pattern of per-tenant
 * depth thresholds.
 */

import type { QueueAdapter } from './adapter.js'

/**
 * Default backpressure threshold per tenant. 1000 pending items
 * signals the system should pause accepting new work.
 */
const DEFAULT_MAX_QUEUE_DEPTH = 1000

/**
 * Evaluates whether the queue has exceeded its capacity threshold.
 * Producers should call isBackpressured before enqueuing and skip
 * the enqueue when it returns true.
 *
 * Time: O(1) for cached reads, O(1) for check (single adapter call).
 * Space: O(1).
 */
export type BackpressureChecker = {
  /**
   * Queries the adapter for current pending job count and updates the
   * cached state. The brainId scopes the depth check; pass an empty
   * string for global depth.
   */
  check(brainId: string): Promise<boolean>

  /**
   * Returns the last known backpressure state without querying the
   * adapter. Safe to call from hot paths where a stale value is
   * acceptable.
   */
  isBackpressured(): boolean

  /**
   * Returns the configured threshold.
   */
  maxDepth(): number

  /**
   * Returns the last observed queue depth from the most recent check
   * call. Returns 0 before the first check.
   */
  lastDepth(): number
}

/**
 * Creates a backpressure checker bound to the given adapter.
 * A maxQueueDepth of zero or negative falls back to
 * DEFAULT_MAX_QUEUE_DEPTH (1000).
 */
export const createBackpressureChecker = (
  adapter: QueueAdapter,
  maxQueueDepth: number,
): BackpressureChecker => {
  const safeMax = maxQueueDepth > 0 ? maxQueueDepth : DEFAULT_MAX_QUEUE_DEPTH
  let pressured = false
  let observedDepth = 0

  return {
    async check(brainId: string): Promise<boolean> {
      const counts = await adapter.countByStatus(brainId === '' ? undefined : brainId)
      const depth = counts.pending ?? 0
      observedDepth = depth
      pressured = depth >= safeMax
      return pressured
    },

    isBackpressured(): boolean {
      return pressured
    },

    maxDepth(): number {
      return safeMax
    },

    lastDepth(): number {
      return observedDepth
    },
  }
}
